File: RxJava3/app/src/main/java/com/nghianv/rxjava3/MainActivity.java
package com.nghianv.rxjava3;

import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.internal.subscribers.StrictSubscriber;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity implements Api {
	private final static String TAG = "MainActivity";
	private Disposable disposable;


	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		//
		//loadAndSubscribe();
		//loadAndSurcribe2();
		loadAndSurcribeFootballPlayer();
	}


	private void loadAndSurcribeFootballPlayer() {
		Observable<String> footballPlayersObservable = Observable.just("Messi", "Ronaldo", "Modric", "Manadona", "Roney");
		footballPlayersObservable
				.subscribeOn(Schedulers.io())
				.filter(new Predicate<String>() {
					@Override
					public boolean test(@NotNull String s) throws Exception {
						return s.toLowerCase().startsWith("m");
					}
				})
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(new Observer<String>() {
					@Override
					public void onSubscribe(Disposable d) {
						disposable = d;
						Log.d(TAG, "onSubscribe");
					}

					@Override
					public void onNext(String s) {
						Log.d(TAG, "Name: " + s);
					}

					@Override
					public void onError(Throwable e) {
						Log.d(TAG, "onError: " + e.getMessage());
					}

					@Override
					public void onComplete() {
						Log.d(TAG, "onComplete: All FootballPlayer are emmited");
					}
				});

	}

	@SuppressLint("CheckResult")
	private void loadAndSurcribe2() {
		Observable.just("VietNam", "England", "US", "Canana", "France", "Japan")
				.subscribeOn(Schedulers.io())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(new Observer<String>() {
					           @Override
					           public void onSubscribe(Disposable d) {
						           disposable = d;
						           Log.d(TAG, "onSubscribe");
					           }

					           @Override
					           public void onNext(String s) {
						           Log.d(TAG, "Country2 >>: " + s);
					           }

					           @Override
					           public void onError(Throwable e) {
						           Log.d(TAG, "onError: " + e.getMessage());
					           }

					           @Override
					           public void onComplete() {
						           Log.d(TAG, "onComplete: All Countries are emitted");
					           }
				           }
				);
	}


	private void loadAndSubscribe() {
		Observable<String> countryObservable = Observable.just("Africa", "Belgium", "Canada", "Denmark", "France");
		Observer<String> animalsObserver = getCountriesObserver();
		countryObservable
				.subscribeOn(Schedulers.io())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(animalsObserver);
	}

	private Observer<String> getCountriesObserver() {
		return new Observer<String>() {
			@Override
			public void onSubscribe(Disposable d) {
				Log.d("MainActivity", "onSubscribe");
			}

			@Override
			public void onNext(String s) {
				Log.d("MainActivity", "Country >>: " + s);
			}

			@Override
			public void onError(Throwable e) {
				Log.e("MainActivity", "onError: " + e.getMessage());
			}

			@Override
			public void onComplete() {
				Log.d("MainActivity", "All Countries are emitted!");
			}
		};
	}

	@Override
	public Observable<String> searchMovie(String query) {
		return null;
	}

	@Override
	public List<Movie> parse(String json) {
		return null;
	}

	@SuppressLint("CheckResult")
	public void printChar(String s) {
		Observable.range(0, 9)
				.filter(it -> {
					return it % 2 == 1;
				})
				.map(it -> it.toString())
				.subscribeOn(Schedulers.io())
				.observeOn(AndroidSchedulers.mainThread())
				.subscribe(string -> printChar(string));

	}

	@SuppressLint("CheckResult")
	private void oparatorMap() {
		Observable.just("Hello, world!")
				.map(
						s -> s + "-Dan"
				).subscribe(System.out::println).dispose();
		Observable.just("Hello!")
				.map(
				String::hashCode
		).subscribe(System.out::println);
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		// don't send events once the activity is destroy
		disposable.dispose();
	}
}

File: RxJava4/app/src/main/java/com/example/rxjava4/MainActivity.kt
package com.example.rxjava4

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import io.reactivex.Flowable
import io.reactivex.FlowableSubscriber
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.rxkotlin.toObservable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //
        buttonClick.setOnClickListener({ startStream2() })
        demoContcat();
    }

    @SuppressLint("CheckResult")
    private fun demoContcat() {
        val data: MutableList<Int> = mutableListOf()
        Flowable.concat(Flowable.just(getDataLocal(), getDataServer()))
            .subscribe(
                { list ->
                    for (item in list)
                    if (!data.contains(item)) data.add(item)
                },
                { error -> Log.i(TAG, "onError: ${error.printStackTrace()}") },
                {
                    for (item in data) {
                        Log.i(TAG, "onComplete: ${item}")
                    }

                }
            )
    }
    private fun getDataLocal(): Flowable<List<Int>> {
        val data: MutableList<Int> = ArrayList()
        data.add(1)
        data.add(2)
        data.add(3)
        return Flowable.just(data)
    }

    private fun getDataServer(): Flowable<List<Int>> {
        val data: MutableList<Int> = ArrayList()
        data.add(3)
        data.add(4)
        data.add(5)
        return Flowable.just(data)
    }


    private fun startStream() {
        //Create an Observable
        val myObservable = getObservable()

        //Create an Observer
        val myObserver = getObserver()

        // Subscribe myObserver to Observable
        myObservable.subscribe(myObserver)
    }

    @SuppressLint("CheckResult")
    private fun startStream2() {
        val list = listOf("1", "2", "3", "4", "5")

        Observable.fromIterable(list)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { println("ReceivedList: ${it}") },
                { println("Error: ${it.printStackTrace()}") },
                { println("onComplete Subcribe list") }
            )


        list.toObservable()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { println(it) },
                { println("Error: ${it.printStackTrace()}") },
                { println("onComplete") }
            )

        list.toObservable()
            .subscribeBy(
                onNext = { println(it) },
                onError = { it.printStackTrace() },
                onComplete = { println("onComplete!") }

            )
    }

    private fun getObserver(): Observer<String> {
        return object : Observer<String> {
            override fun onSubscribe(d: Disposable) {}

            override fun onNext(t: String) {
                Log.d(TAG, "onNext: $t")
            }

            override fun onError(e: Throwable) {
                Log.d(TAG, "onError: ${e.message}")
            }

            override fun onComplete() {
                Log.d(TAG, "onComplete")
            }

        }
    }

    private fun getObservable(): Observable<String> {
        return Observable.just("1", "2", "3", "4", "5")
    }

}